### PR TITLE
til::size

### PR DESCRIPTION
**Pull request by miniksa**
## Summary of the Pull Request
Introduces convenience type `til::size` which automatically implements our best practices for size-related types and provides automatic conversions in/out of the relevant types.

## PR Checklist
* [x] In support of Differental Rendering #778
* [X] I work here.
* [x] Tests added/passed
* [x] I'm a core contributor.

## Detailed Description of the Pull Request / Additional comments
- Automatically converts in from anything with a X/Y (console `COORD`) or cx/cy (Win32 `SIZE`)
- Automatically converts out to `COORD`, `SIZE`, or `D2D1_SIZE_F`.
- Constructs from bare integers written into source file
- Default constructs to empty
- Uses Chromium Math for all basic math operations (+, -, *, /)
- Provides equality tests
- Adds initial proposal for division-to-ceiling (round up division) that attempts to `ceil` without any floating point math.
- Accessors for height/width
- Type converting accessors (that use safe conversions and throw) for height/width
- Convenience function for area calculation (as that's common with type) and uses safe math to do it.
- TAEF/WEX Output and Comparators so they will print very nicely with `VERIFY` and `Log` macros in our testing suite.

## Validation Steps Performed
- See automated tests of functionality.

### CODE REVIEW

File: src/inc/til/size.h [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    class size
+    {
+    public:
+        constexpr size() noexcept :
+            size(0, 0)
+        {
+
+        }
+
+#if defined(_M_AMD64) || defined(_M_ARM64)
```
Comment (DHowett-MSFT) [inline src/inc/til/size.h]: doc why

File: src/inc/til/size.h [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        size divide_ceil(const size& other) const
+        {
+            // Divide normally to get the floor.
+            const size floor = *this / other;
+
+            ptrdiff_t adjWidth = 0;
+            ptrdiff_t adjHeight = 0;
+
+            // Check for width remainder, anything not 0.
+            if (_width % other._width)
```
Comment (german-one) [inline src/inc/til/size.h]: May I suggest to use multiplication with `floor._width` for performance reasons? Modulus invokes another slow division that has been already performed to get `floor`.

**Comment by miniksa** [comment]
Nnnnn I don't think I checked SA here and it's not used in an SA-enabled project yet. I should do that.

File: src/inc/til/size.h (hunk 2 above)
Comment (miniksa) [reply src/inc/til/size.h]: I'm all for performance. Please, do suggest.

But... Sorry, what are you saying? That I should multiply by the already floored number and subtract? Or is there something trickier here that I'm not seeing?

File: src/inc/til/size.h (hunk 1 above)
Comment (miniksa) [reply src/inc/til/size.h]: done in next push

File: src/inc/til/size.h (hunk 2 above)
Comment (german-one) [reply src/inc/til/size.h]: You performed division:
`_width / other._width = floor._width`
If you turn it around then should `other._width * floor._width` be equal with `_width` for a division with remainder 0.

So, this should do the trick:
`if (other._width * floor._width != _width)`
For `_height`, respectively.

Comment (miniksa) [reply src/inc/til/size.h]: Oh with the comparison. Yes. Yes. OK.

**Comment by miniksa** [comment]
OH. I should put a natvis on this too, eh, @DHowett-MSFT?

**Comment by DHowett-MSFT** [comment]
i do like me some good natvisses

File: src/inc/til/size.h [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(height).AssignIfValid(&_height));
+        }
+
+        constexpr size(ptrdiff_t width, ptrdiff_t height) noexcept :
+            _width(width),
+            _height(height)
+        {
+        }
+
+        template<typename TOther>
```
Comment (DHowett-MSFT) [inline src/inc/til/size.h]: nit: minor docs; "this is for converting from something with X, Y"

**Comment by carlos-zamora** [comment]
Is there a way I use this for clamped math? Would I just have to catch the HRESULT after each operation? It'd be nice if I could just say "always clamp to these dimensions" (i.e. text buffer size), and it would "just work". But that might also be outside the scope of this PR. Thoughts?

**Comment by miniksa** [comment]
>     1. Is this possible to use in a clamped context?
No, not yet. I didn't need that yet... so I didn't put it in.
This isn't really intended to be the END of writing `til::size`. The beginning and what I needed for rendering. I just didn't want to go off the deep end and be writing the support class forever so I did just about as much as I needed then sent it in.

>     2. Should this maybe be `til::vec2` that exposes both `width()` and `x()` (for example)?
What you're not seeing yet is that `til::point` is coming for x/y. And they don't inherit or whatever from each other because I wanted the distinction for `til::rectangle` to construct with either `point/point` (as the two corners) or `point/size` (as the origin and the dimensions).

**Comment by miniksa** [comment]
> Is there a way I use this for clamped math? Would I just have to catch the HRESULT after each operation? It'd be nice if I could just say "always clamp to these dimensions" (i.e. text buffer size), and it would "just work". But that might also be outside the scope of this PR. Thoughts?

I don't have a clamped one yet because I didn't have a specific use case in my mind yet and I was trying not to write this class forever. If we have a use for clamped, then we can add it. This isn't the end of writing `til::size`. We'll add more stuff to it as we need it.

Maybe it takes an optional bool parameter on the constructors that sets it to clamped mode instead of checked mode? We can figure that out in a future revision, IMO.

File: src/inc/til/size.h (hunk 3 above)
Comment (miniksa) [reply src/inc/til/size.h]: done

**Pull request merged**